### DISCUSSION TITLE
Swapping feature implementation options

### DISCUSSION THREAD

**Discussion by bjornorri** [rxhanson/Rectangle#833] [Ideas]
Hi @rxhanson!

I would like to implement a mechanism for swapping windows between displays, similar to what was discussed in #819.

My desired use case:

1. Display A contains maximized editor and display B contains browser on left half and email on right half.
2. Shortcut is triggered.
3. Display A contains browser on left half and email on right half. Display B contains maximized editor.

I started looking into implementing this and realized there are several different implementations that could accomplish my use case. I'm not sure which one makes the most sense as a Rectangle feature, so I'd love to get your opinion.

# Implementation Options
## Option 1 - Swap all windows between displays
This option would move all windows from display A to display B and vice-versa.

### Implementation
The [allWindowsOnScreen](https://github.com/rxhanson/Rectangle/blob/fc1beabd440bc44156a49690fb630562a8348415/Rectangle/MultiWindow/MultiWindowManager.swift#L30-L65) function can be used to select the windows on display A and display B. They can then be moved one-by-one.

### Pros
* Simple implementation.
* Triggered by a single shortcut.

### Cons
* Not really the behaviour I want. I might e.g. want to swap my maximized browser and editor but still keep email, chat, etc. on my secondary screen.

## Option 2 - Swap visible windows between displays
This option would move the currently visible windows from display A to display B and vice-versa.

### Implementation
I'm not sure what the best way is to select all visible windows on a given screen. One idea would be to use `AXUIElementCopyElementAtPosition` to sample locations all over the screen and pick out the corresponding windows. They would then be moved one-by-one like in option 1.

### Pros
* Kind of simple implementation.
* Triggered by a single shortcut.
* More intuitive behaviour than option 1 (IMO).

### Cons
* Location sampling is not the cleanest solution.

## Option 3 - Add swap flag that works with any window action
Let's imagine we're going to trigger a window action that moves the window from rect A to rect B. This option would add a mechanism to tell Rectangle to not only perform that action but also to move all windows currently in rect B to rect A. This mechanism could e.g. be triggered by a shortcut that you trigger before your main window action shortcut.

### Implementation
Not sure how this would be implemented. A swap shortcut would probably activate some flag that then needs to be integrated into other window actions.

### Pros
* Flexibility. This could be used to swap halves, thirds, etc. as well as maximized display windows.

### Cons
* This would be triggered with two shortcuts. One to activate swapping and one for the window action.
* More complex than the other options since this is a more "global feature".

# Questions
1. Do any of these options seem like a reasonable feature to you, which would be your preference?
2. Do you know if moving multiple windows can scramble their z-ordering, depending on the order they're moved in?

**Comment by rxhanson** [comment]
Thanks for looking to contribute, and for thinking through these options. At first cut, I only allow features like this to be toggled via terminal command. This gives us flexibility to try different things out, and doesn't add any complexity to the UI.

> 1. Do any of these options seem like a reasonable feature to you, which would be your preference?

The best option is the one that you have desire to implement and use. I'd prefer to just let you build the thing that you desire the most, but if you want me to pick a path or if you're really curious then I can give you my opinion.

> 2. Do you know if moving multiple windows can scramble their z-ordering, depending on the order they're moved in?

Z-ordering is not affected. The only way to change Z-ordering is to bring a window frontmost, if desired.



I'm certain that there are plenty of edge cases to figure out with this feature, but one that jumps out at me is what happens if the user has more than 2 displays? (likely this feature would be disabled for only 1 display?)


On the implementation for option 3, you'd likely be looking to use the `WindowHistory` to find windows that have been previously snapped to certain positions. Adding things to all window actions is typically undesirable.

**Reply by bjornorri** [reply]
Thanks for the quick feedback.

    ...what happens if the user has more than 2 displays? (likely this feature would be disabled for only 1 display?)
    
I was thinking that this feature would provide `swapPreviousDisplay` and `swapNextDisplay` actions that operate relative to the display of the currently active window (like `previousDisplay` and `nextDisplay`). Nothing would happen with a single display.

I will experiment a bit and see which of these options is most promising.

**Reply by rxhanson** [reply]
> I was thinking that this feature would provide swapPreviousDisplay and swapNextDisplay actions that operate relative to the display of the currently active window (like previousDisplay and nextDisplay). Nothing would happen with a single display.

Ah, that sounds good to me.

**Comment by andrecarlucci** [comment]
Really liked the idea!
